File: src/main/java/nl/chimpgamer/networkmanager/extensions/discordbot/DiscordBot.java
package nl.chimpgamer.networkmanager.extensions.discordbot;

import lombok.Getter;
import net.dv8tion.jda.api.entities.Guild;
import nl.chimpgamer.networkmanager.api.NMListener;
import nl.chimpgamer.networkmanager.api.extensions.NMExtension;
import nl.chimpgamer.networkmanager.api.utils.PlatformType;
import nl.chimpgamer.networkmanager.bungeecord.NetworkManager;
import nl.chimpgamer.networkmanager.bungeecord.commands.NMCommandManager;
import nl.chimpgamer.networkmanager.extensions.discordbot.api.models.Token;
import nl.chimpgamer.networkmanager.extensions.discordbot.commands.mc.*;
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.CommandSetting;
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.CommandSettings;
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.Messages;
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.Settings;
import nl.chimpgamer.networkmanager.extensions.discordbot.listeners.NetworkManagerListeners;
import nl.chimpgamer.networkmanager.extensions.discordbot.listeners.bungee.ChatListener;
import nl.chimpgamer.networkmanager.extensions.discordbot.listeners.bungee.JoinLeaveListener;
import nl.chimpgamer.networkmanager.extensions.discordbot.listeners.bungee.RedisBungeeListener;
import nl.chimpgamer.networkmanager.extensions.discordbot.manager.DiscordManager;
import nl.chimpgamer.networkmanager.extensions.discordbot.manager.DiscordUserManager;
import nl.chimpgamer.networkmanager.extensions.discordbot.tasks.TokenExpiryTask;
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.DependencyDownloader;
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.DiscordPlaceholders;
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.MySQL;
import org.checkerframework.checker.nullness.qual.NonNull;

import java.util.ArrayList;
import java.util.List;

@Getter
public final class DiscordBot extends NMExtension {
    private static DiscordBot instance;

    private List<Integer> tasks;
    private List<NMListener> listeners;

    // Configuration files
    private Settings settings;
    private CommandSettings commandSettings;
    private Messages messages;

    /*private ConfigManager configManager;
    private MessagesConfigManager messagesConfigManager;*/
    private MySQL mySQL;
    private DiscordUserManager discordUserManager;
    private DiscordManager discordManager;

    public static DiscordBot getInstance() {
        return instance;
    }

    @Override
    public void onEnable() {
        // Extension startup logic
        instance = this;

        if (this.getNetworkManager().getPlatformType() != PlatformType.BUNGEECORD) {
            this.getLogger().severe("Hey, this NetworkManager extension is for BungeeCord only!");
            return;
        }
        DependencyDownloader dd = new DependencyDownloader(this);
        dd.downloadDependency(
                "https://github.com/DV8FromTheWorld/JDA/releases/download/v4.0.0/JDA-4.0.0_39-withDependencies-no-opus.jar",
                "JDA",
                "JDA-4.0.0_39-withDependencies-no-opus");

        this.tasks = new ArrayList<>();
        this.listeners = new ArrayList<>();

        // Initialize configuration files
        this.initSettings();
        this.initCommands();
        this.initMessages();

        this.mySQL = new MySQL(this);
        this.discordUserManager = new DiscordUserManager(this);
        this.discordUserManager.load();

        this.discordManager = new DiscordManager(this);
        if (!this.discordManager.init()) {
            this.disable();
            return;
        }

        this.registerCommands();
        this.registerListeners();

        if (this.getNetworkManager().isRedisBungee()) {
            this.getNetworkManager().registerListener(new RedisBungeeListener(this));
            this.getNetworkManager().getRedisBungee().registerPubSubChannels("NetworkManagerDiscordBot");
        }

        this.getNetworkManager().getPlaceholderManager().registerPlaceholder(new DiscordPlaceholders(this));
    }

    @Override
    public void onDisable() {
        // Extension shutdown logic
        this.expireTokens();
        this.unregisterListeners();
        this.getNetworkManager().getCommandManager().unregisterAllBySource(this.getInfo().getName());
        this.getDiscordManager().shutdownJDA();
    }

    @Override
    protected void onConfigsReload() {
        this.getSettings().reload();
        this.getCommandSettings().reload();
        this.getMessages().reload();
    }

    private void initSettings() {
        this.settings = new Settings(this);
        this.settings.init();
    }

    private void initCommands() {
        this.commandSettings = new CommandSettings(this);
        this.commandSettings.init();
    }

    private void initMessages() {
        this.messages = new Messages(this);
        this.messages.init();
    }

    private void registerListeners() {
        this.getListeners().add(new NetworkManagerListeners(this));
        for (NMListener listener : this.getListeners()) {
            this.getEventHandler().registerListener(listener);
        }
        this.getNetworkManager().registerListeners(
                new JoinLeaveListener(this),
                new ChatListener(this)
        );
    }

    private void registerCommands() {
        NMCommandManager commandManager = this.getNetworkManager().getCommandManager();
        if (CommandSetting.MINECRAFT_BUG_ENABLED.getAsBoolean()) {
            commandManager.registerCommand(this.getInfo().getName(), new BugCommand(this, CommandSetting.MINECRAFT_BUG_COMMAND.getAsString()));
        }
        if (CommandSetting.MINECRAFT_SUGGESTION_ENABLED.getAsBoolean()) {
            commandManager.registerCommand(this.getInfo().getName(), new SuggestionCommand(this, CommandSetting.MINECRAFT_SUGGESTION_COMMAND.getAsString()));
        }
        if (CommandSetting.MINECRAFT_DISCORD_ENABLED.getAsBoolean()) {
            commandManager.registerCommand(this.getInfo().getName(), new DiscordCommand(this, "discord"));
        }
        commandManager.registerCommands(this.getInfo().getName(),
                new VerifyCommand(this, CommandSetting.MINECRAFT_VERIFY_COMMAND.getAsString()),
                new NetworkManagerBotCommand(this, "networkmanagerbot")
        );
    }

    private void unregisterListeners() {
        for (NMListener listener : this.getListeners()) {
            this.getEventHandler().unregisterListener(listener);
        }
    }

    public void sendRedisBungee(String message) {
        this.getScheduler().runAsync(() -> this.getNetworkManager().getRedisBungee().sendChannelMessage("NetworkManagerDiscordBot", message), false);
    }

    private void expireTokens() {
        if (this.getDiscordUserManager() != null) {
            for (Token token : this.getDiscordUserManager().getTokens()) {
                this.getScheduler().runSync(new TokenExpiryTask(this, token));
            }
        }
    }

    public @NonNull Guild getGuild() {
        return this.getDiscordManager().getGuild();
    }

    @Override
    public NetworkManager getNetworkManager() {
        return (NetworkManager) super.getNetworkManager();
    }
}

File: src/main/kotlin/nl/chimpgamer/networkmanager/extensions/discordbot/DiscordBot.kt
package nl.chimpgamer.networkmanager.extensions.discordbot

import io.github.slimjar.app.builder.ApplicationBuilder
import net.dv8tion.jda.api.entities.Guild
import nl.chimpgamer.networkmanager.api.extensions.NMExtension
import nl.chimpgamer.networkmanager.api.utils.PlatformType
import nl.chimpgamer.networkmanager.common_proxy.plugin.NetworkManagerPluginProxyBase
import nl.chimpgamer.networkmanager.extensions.discordbot.commands.mc.*
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.CommandSetting
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.CommandSettings
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.Messages
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.Settings
import nl.chimpgamer.networkmanager.extensions.discordbot.listeners.NetworkManagerListeners
import nl.chimpgamer.networkmanager.extensions.discordbot.listeners.bungee.BungeeCordJoinLeaveListener
import nl.chimpgamer.networkmanager.extensions.discordbot.listeners.bungee.RedisBungeeListener
import nl.chimpgamer.networkmanager.extensions.discordbot.listeners.velocity.VelocityJoinLeaveListener
import nl.chimpgamer.networkmanager.extensions.discordbot.manager.DiscordManager
import nl.chimpgamer.networkmanager.extensions.discordbot.manager.DiscordUserManager
import nl.chimpgamer.networkmanager.extensions.discordbot.tasks.ActivityUpdateTask
import nl.chimpgamer.networkmanager.extensions.discordbot.tasks.TokenExpiryTask
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.DiscordPlaceholders
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.MySQL
import java.io.File
import java.time.Duration
import java.time.Instant
import java.util.logging.Level

class DiscordBot : NMExtension() {

    // Configuration files
    val settings = Settings(this)
    val commandSettings = CommandSettings(this)
    lateinit var messages: Messages

    val mySQL = MySQL(this)
    val discordUserManager = DiscordUserManager(this)
    lateinit var discordManager: DiscordManager

    private val activityUpdateTask = ActivityUpdateTask(this)

    public override fun onEnable() { // Extension startup logic
        val dependencyDirectory = File(networkManagerPlugin.dataFolder, "libraries")
        logger.log(Level.INFO, "Loading Libraries...")
        logger.log(Level.INFO, "Note: This might take a few minutes on first run. Kindly ensure internet connectivity.")
        val startInstant = Instant.now()
        try {
            val slimJarJsonUrl = javaClass.classLoader.getResource("slimjar.json")
            ApplicationBuilder
                .appending("DiscordBot")
                .dependencyFileUrl(slimJarJsonUrl)
                .downloadDirectoryPath(dependencyDirectory.toPath())
                .logger { s, anies -> logger.log(Level.INFO, s, anies) }
                .build()
            val endInstant = Instant.now()
            val timeTaken = Duration.between(startInstant, endInstant).toMillis()
            val timeTakenSeconds = timeTaken / 1000.0
            logger.log(Level.INFO, "Loaded libraries in {0} seconds", timeTakenSeconds)
        } catch (exception: Exception) {
            logger.log(
                Level.SEVERE,
                "Unable to load dependencies... Please ensure an active Internet connection on first run!"
            )
            exception.printStackTrace()
            disable()
            return
        }

        instance = this
        if (networkManager.platformType.isProxy.not()) {
            logger.severe("Hey, this NetworkManager extension is for BungeeCord and Velocity only!")
            return
        }

        // Initialize MySQL. If False stop plugin.
        if (mySQL.initialize().not()) {
            disable()
            return
        }

        // Initialize configuration files
        settings.init()
        commandSettings.init()
        messages = Messages(this)
        messages.init()

        discordUserManager.load()

        discordManager = DiscordManager(this)
        if (!discordManager.init()) {
            disable()
            return
        }

        registerCommands()
        registerListeners()
        activityUpdateTask.start()
        if (networkManager.isRedisBungee) {
            networkManager.registerListener(RedisBungeeListener(this))
        }
        networkManager.placeholderManager.registerPlaceholder(DiscordPlaceholders(this))
    }

    public override fun onDisable() { // Extension shutdown logic
        expireTokens()
        activityUpdateTask.stop()

        this.discordManager.shutdownJDA()
    }

    override fun onConfigsReload() {
        this.settings.reload()
        this.commandSettings.reload()
        this.messages.reload()
    }

    private fun registerListeners() {
        NetworkManagerListeners(this)
        if (networkManager.platformType === PlatformType.BUNGEECORD) {
            networkManager.registerListeners(
                BungeeCordJoinLeaveListener(this)
            )
        } else if (networkManager.platformType === PlatformType.VELOCITY) {
            networkManager.registerListeners(
                VelocityJoinLeaveListener(this)
            )
        }
    }

    private fun registerCommands() {
        val cloudCommandManager = networkManager.cloudCommandManager

        if (commandSettings.getBoolean(CommandSetting.MINECRAFT_BUG_ENABLED)) {
            cloudCommandManager.commandManager.command(CloudBugCommand(this)
                .getCommand(commandSettings.getString(CommandSetting.MINECRAFT_BUG_COMMAND)))
        }
        if (commandSettings.getBoolean(CommandSetting.MINECRAFT_SUGGESTION_ENABLED)) {
            cloudCommandManager.commandManager.command(CloudSuggestionCommand(this)
                .getCommand(commandSettings.getString(CommandSetting.MINECRAFT_SUGGESTION_COMMAND)))
        }
        if (commandSettings.getBoolean(CommandSetting.MINECRAFT_DISCORD_ENABLED)) {
            cloudCommandManager.commandManager.command(CloudDiscordCommand(this).getCommand("discord"))
        }

        cloudCommandManager.commandManager.command(CloudRegisterCommand(this)
            .getCommand(commandSettings.getString(CommandSetting.MINECRAFT_REGISTER_COMMAND), *commandSettings.getString(CommandSetting.MINECRAFT_REGISTER_ALIASES).split(", ").toTypedArray()))
        cloudCommandManager.commandManager.command(CloudUnregisterCommand(this)
            .getCommand(commandSettings.getString(CommandSetting.MINECRAFT_UNREGISTER_COMMAND), *commandSettings.getString(CommandSetting.MINECRAFT_UNREGISTER_ALIASES).split(", ").toTypedArray()))
        cloudCommandManager.annotationParser.parse(CloudNetworkManagerBotCommand(this))
    }

    private fun expireTokens() {
        for (token in this.discordUserManager.tokens) {
            this.scheduler.runSync(TokenExpiryTask(this, token))
        }
    }

    val guild: Guild
        get() = this.discordManager.guild

    override val networkManager: NetworkManagerPluginProxyBase
        get() = super.networkManager as NetworkManagerPluginProxyBase

    fun isDiscordManagerInitialized(): Boolean = this::discordManager.isInitialized

    companion object {
        @JvmStatic
        lateinit var instance: DiscordBot
    }
}